### ISSUE TITLE
Support for borderless and transparent windows

### ISSUE THREAD

**Issue by pcwalton** [andlabs/libui#79]
For http://github.com/servo/servo we draw the entire window ourselves with OpenGL, so that we can perform global optimizations to reduce GPU state changes and so forth. This means that we need the ability to create a window without a title bar. It also, somewhat surprisingly, means that we need a transparent window, because on Mac OS X windows have rounded corners and the client is responsible for drawing this. Generally, windowing systems have to be explicitly informed that windows are to be opaque in order to disable occlusion culling.

There are a number of ways we could do this: we could have a `uiNewBorderlessWindow(const char *title, int width, int height, int hasMenubar, int transparent)` call, or we could create a `uiWindowOptions` type of structure and refactor the `uiNewWindow` call to use that. Any thoughts?

**Comment by andlabs** [comment]
So would this be:

- no WM_CAPTION or WM_THICKFRAME on Windows, or should I go all the way and define the non-client area myself?
- NSBorderlessWindowMask is sufficient, or would I need to do more there?
- would an undecorated window be enough on Linux?

And how would you be handling moving and resizing of these windows?

**Comment by pcwalton** [comment]
I think all of your suggestions seem OK. I don't think we'd need a custom client area, though I'm less knowledgeable about Win32.

In my case I'll handle the window moving myself via drag events on a `uiOpenGLArea`, but I do need the user to be able to resize the window by dragging the edges. (I can't speak for others though; apps using this functionality to display splash screens, for example, might not want their borderless windows to be resizable.)

**Comment by andlabs** [comment]
Right; the question is how much I would need to write myself, how much the OS provides, and what new APIs I would need on top of just adding the flag (for instance, an API that repositions a window)

**Comment by pcwalton** [comment]
I think all we need is the flags you mentioned in addition to an API to move the window. (BTW, I'm happy to do the implementation work here.)

**Comment by andlabs** [comment]
Position manipulation is now in. Next is sizing.

I wonder if I can make a window borderless after the fact on OS X...

**Comment by andlabs** [comment]
Sizes added.

**Comment by andlabs** [comment]
I added calls to make an existing window borderless. What do you think?

I've forgotten what you mean by transparent windows; what did you have in mind there?

**Comment by pcwalton** [comment]
Actually, I don't need transparent windows after all, so don't worry about it. :)

**Comment by andlabs** [comment]
Erm, what does Servo do on Wayland about window positions and sizes?

**Comment by PaulBGD** [comment]
For transparent windows, I'm guessing he's talking about those launcher windows for programs that are partially transparent.

**Comment by pcwalton** [comment]
I'm not quite sure what you mean—what is special on Wayland here?

**Comment by andlabs** [comment]
Wayland doesn't allow you to specify an exact position or size for a window. Instead, it gives you an API that you can use to initiate a user drag or resize operation.

**Comment by PaulBGD** [comment]
Could you use those APIs to replicate the behavior on other desktops? Like make a resize operation from x to y? I'm curious how to support Wayland when position/sizing is so different to other desktops.

**Comment by andlabs** [comment]
How does this API look to you?

```diff
diff --git a/ui.h b/ui.h
index 5cf01b5..518d366 100644
--- a/ui.h
+++ b/ui.h
@@ -289,6 +289,22 @@ struct uiAreaHandler {
 	int (*KeyEvent)(uiAreaHandler *, uiArea *, uiAreaKeyEvent *);
 };
 
+// TODO RTL layouts?
+// TODO reconcile edge and corner naming
+_UI_ENUM(uiWindowResizeEdge) {
+	uiWindowResizeEdgeLeft,
+	uiWindowResizeEdgeTop,
+	uiWindowResizeEdgeRight,
+	uiWindowResizeEdgeBottom,
+	uiWindowResizeEdgeTopLeft,
+	uiWindowResizeEdgeTopRight,
+	uiWindowResizeEdgeBottomLeft,
+	uiWindowResizeEdgeBottomRight,
+	// TODO have one for keyboard resizes?
+	// TODO GDK doesn't seem to have any others, including for keyboards...
+	// TODO way to bring up the system menu instead?
+};
+
 #define uiArea(this) ((uiArea *) (this))
 // TODO give a better name
 // TODO document the types of width and height
@@ -296,6 +312,11 @@ _UI_EXTERN void uiAreaSetSize(uiArea *a, int width, int height);
 // TODO uiAreaQueueRedraw()
 _UI_EXTERN void uiAreaQueueRedrawAll(uiArea *a);
 _UI_EXTERN void uiAreaScrollTo(uiArea *a, double x, double y, double width, double height);
+// TODO document these can only be called within Mouse() handlers
+// TODO should these be allowed on scrolling areas?
+// TODO decide which mouse events should be accepted; Down is the only one guaranteed to work right now
+_UI_EXTERN void uiAreaBeginUserWindowMove(uiArea *a);
+_UI_EXTERN void uiAreaBeginUserWindowResize(uiArea *a, uiWindowResizeEdge edge);
 _UI_EXTERN uiArea *uiNewArea(uiAreaHandler *ah);
 _UI_EXTERN uiArea *uiNewScrollingArea(uiAreaHandler *ah, int width, int height);
 ```

Mind the TODO noise.

**Comment by andlabs** [comment]
They are present on the other platforms, and have been for a while. Any updates on its satisfactory-ness?
